eth/db/accesslog.py

```diff
@@ -0,0 +1,106 @@
+from contextlib import contextmanager
+import logging
+from typing import (
+    Iterator,
+    Set,
+)
+
+from eth.abc import (
+    AtomicWriteBatchAPI,
+    AtomicDatabaseAPI,
+    DatabaseAPI,
+)
+from eth.db.backends.base import (
+    BaseDB,
+)
+from eth.db.atomic import (
+    BaseAtomicDB,
+)
+
+
+class KeyAccessLoggerDB(BaseDB):
+    """
+    Wraps around a database, and tracks all the keys that were read since initialization.
+    """
+
+    logger = logging.getLogger("eth.db.KeyAccessLoggerDB")
+
+    def __init__(self, wrapped_db: DatabaseAPI, log_missing_keys: bool=True) -> None:
+        """
+        :param log_missing_keys: True if a key is added to :attr:`keys_read` even if the
+            key/value does not exist in the database.
+        """
+        self.wrapped_db = wrapped_db
+        self._keys_read: Set[bytes] = set()
+        self._log_missing_keys = log_missing_keys
+
+    @property
+    def keys_read(self) -> Set[bytes]:
+        # Make a defensive copy so callers can't modify the list externally
+        return set(self._keys_read)
```

Comment:
Returning a new `set` provides the protection we need but maybe that's a good candidate to use `frozenset` just to drive home the point that this is meant to be readonly.